devtools/gradle/src/main/java/io/quarkus/gradle/dependency/ExtensionDependency.java

```diff
@@ -0,0 +1,101 @@
+    Dependency dependencyCondition;
+
+    ExtensionDependency(ModuleVersionIdentifier extensionId, List<Dependency> conditionalDependencies,
+            Dependency dependencyCondition) {
+        this.extensionId = extensionId;
+        this.conditionalDependencies = conditionalDependencies;
+        this.dependencyCondition = dependencyCondition;
+    }
+
+    public boolean needsResolution(Set<ResolvedArtifact> resolvedArtifacts) {
+        for (Dependency dependency : conditionalDependencies) {
+            if (!DependencyUtils.exists(resolvedArtifacts, dependency)) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    public void createFeatureVariant(DependencyHandler dependencies, ModuleVersionIdentifier capability) {
+        Dependency dependency = findConditionalDependency(capability);
+
+        if (dependency == null) {
+            throw new GradleException("Trying to add " + capability.getName() + " variant which is not declared by "
+                    + extensionId.getName() + " extension.");
+        }
+
+        dependencies.components(handler -> {
+            handler.withModule(toModuleName(),
+                    componentMetadataDetails -> {
+                        componentMetadataDetails.addVariant(
+                                dependency.getGroup() + ":" + dependency.getName() + ":feature",
+                                variantMetadata -> {
+                                    variantMetadata
+                                            .withDependencies(d -> d.add(DependencyUtils.asDependencyNotation(dependency)));
+                                    variantMetadata.withCapabilities(c -> {
+                                        c.removeCapability(extensionId.getGroup(),
+                                                extensionId.getName());
+                                        c.addCapability(dependency.getGroup(),
+                                                dependency.getName() + DependencyUtils.FEATURE_SUFFIX, dependency.getVersion());
+                                    });
+                                });
+                    });
+        });
+    }
+
+    public void importConditionalDependency(Project project, ModuleVersionIdentifier dependency) {
+        DependencyHandler dependencies = project.getDependencies();
+        createFeatureVariant(dependencies, dependency);
+        ExternalDependency dep = (ExternalDependency) dependencies.create(asDependencyNotation());
+        dep.capabilities(handler -> handler.requireCapability(DependencyUtils.asFeatureName(dependency)));
+
+        Set<ExcludeRule> excludeRules = DependencyUtils.loadExclusionRules(project, asDependencyNotation());
+        for (ExcludeRule excludeRule : excludeRules) {
+            dep.exclude(DependencyUtils.exclusionRuleToMap(excludeRule));
+        }
+
+        project.getDependencies().add(JavaPlugin.IMPLEMENTATION_CONFIGURATION_NAME, dep);
```

Comment:
Conditional dependencies shouldn't be added as direct project dependencies, they should be added as capability dependencies to the extensions that declare conditional dependencies instead. Do you think you can refactor it that way?

Reply:
I don't think we can, each dependency can require one capability (from one I understood), by default, this requires a capability with the same name as the dependency. To require a new capability, we have to declare a new dependency. In the test example, we will end up with something similar too : 

    implementation("org.acme:quarkus-ext-a:1.0-SNAPSHOT")
    implementation("org.acme:quarkus-ext-a:1.0-SNAPSHOT") {
         capabilities {
            requireCapability("org.acme:quarkus-ext-b-feature")
        }
    }

Reply:
I thought we could modify metadata of an existing dependency, couldn't we?

Reply:
we can, that what we do when adding a variant. I will try to require that capability there too.

Reply:
You were right, actually, this can be simplified to updating existing variant and adding a dependency
 to it.